src/utils.ts

```diff
@@ -531,3 +531,81 @@ export async function getFilePathsFromPatterns(
 
   return Array.from(filePaths);
 }
+
+/** Proxy function to use in place of fetch() to bypass CORS restrictions.
+ * It currently doesn't support streaming until this is implemented
+ * https://forum.obsidian.md/t/support-streaming-the-request-and-requesturl-response-body/87381 */
+export async function safeFetch(url: string, options: RequestInit): Promise<Response> {
+  console.log("==================== safeFetch ====================");
+  // Necessary to remove 'content-length' in order to make headers compatible with requestUrl()
+  delete (options.headers as Record<string, string>)["content-length"];
+
+  if (typeof options.body === "string") {
+    const newBody = JSON.parse(options.body ?? {});
+    // frequency_penalty: default 0, but perplexity.ai requires 1 by default.
+    // so, delete this argument for now
+    delete newBody["frequency_penalty"];
+    options.body = JSON.stringify(newBody);
+  }
+
+  const response = await requestUrl({
+    url,
+    contentType: "application/json",
+    headers: options.headers as Record<string, string>,
+    method: "POST",
+    body: options.body?.toString(),
+  });
+
+  return {
+    ok: response.status >= 200 && response.status < 300,
+    status: response.status,
+    statusText: response.status.toString(),
+    headers: new Headers(response.headers),
+    url: url,
+    type: "basic",
+    redirected: false,
+    body: createReadableStreamFromString(response.text),
+    bodyUsed: true,
+    json: () => response.json,
+    text: async () => response.text,
+    clone: () => {
+      throw new Error("not implemented");
+    },
+    arrayBuffer: () => {
+      throw new Error("not implemented");
+    },
+    blob: () => {
+      throw new Error("not implemented");
+    },
+    formData: () => {
+      throw new Error("not implemented");
+    },
+  };
+}
+
+function createReadableStreamFromString(input: string) {
+  return new ReadableStream({
+    start(controller) {
+      // Convert the input string to a Uint8Array
+      const encoder = new TextEncoder();
+      const uint8Array = encoder.encode(input);
+
+      // Push the data to the stream
+      controller.enqueue(uint8Array);
+
+      // Close the stream
+      controller.close();
+    },
+  });
+}
+
+export function err2String(err: any, stack = false) {
+  // maybe to be improved
+  return err instanceof Error
+    ? err.message +
+        "\n" +
+        `${err?.cause ? "more message: " + err.cause.message : ""}` +
```

Comment:
Nit: there's a type error for `err.cause.message`

Reply:
```ts
// type define in lib.es2022.error.d.ts
interface Error {
    cause?: Error;
}
```